[mlir][bufferization] Buffer deallocation: Make op preconditions stricter

matthias-springer opened the pull request:
The buffer deallocation pass checks the IR ("operation preconditions") to make sure that there is no IR that is unsupported. In such a case, the pass signals a failure.

The pass now rejects all ops with unknown memory effects. We do not know whether such an op allocates memory or not. Therefore, the buffer deallocation pass does not know whether a deallocation op should be inserted or not.

Memory effects are queried from the `MemoryEffectOpInterface` interface. Ops that do not implement this interface but have the `RecursiveMemoryEffects` trait do not have any side effects (apart from the ones that their nested ops may have).

Unregistered ops are now rejected by the pass because they do not implement the `MemoryEffectOpInterface` and neither do we know if they have `RecursiveMemoryEffects` or not. All test cases that currently have unregistered ops are updated to use registered ops.

joker-eph commented:
> in particular, we do not know whether an unregistered op allocates memory or not. Therefore, unregistered ops cannot be handled safely in the buffer deallocation pass.

It's a bit concerning because a registered op that does not implement the memory effect interface does not provide more information either.

matthias-springer commented:
That is a problem indeed. I just took a look at `CSE.cpp` to see how ops without the `MemoryEffectOpInterface` are treated. Such ops are assumed to have a side effect.

We currently assume throughout the buffer deallocation infrastructure that an op does not have an allocation side effect if the op does not declare `MemoryEffects::Allocate` via the `MemoryEffectOpInterface`. If an op allocates but does not declare the side effect (or does not implement the interface), the buffer deallocation pass would let the allocation leak.

If an op does not implement `MemoryEffectOpInterface`, we cannot conservatively assume that the op allocates. In that case, we would insert a deallocation for a buffer that was not allocated.

I think we can utilize a bunch of MLIR traits here. Instead of checking whether an op is unregistered or not, something along the lines of:

- Skip all ops that do not operate on buffers (#75126).
- If an op implements `MemoryEffectOpInterface`, we can query that interface directly to see if the op allocates.
- Otherwise, check if the op has a `Pure` or `RecursiveMemoryEffects` trait. In that case, we can safely assume that the op does not have an `Allocate` side effect.
- Otherwise, reject the input IR because we do not have enough information about the op.

Unregistered ops will be rejected automatically because they do not implement any interface or trait. I'm going to give it a try and update this PR.

matthias-springer commented:
Another tricky case would be ops that sometimes allocate and sometimes don't. This is something that we do not support properly at the moment.
```mlir
// With a probability of 40%: %0 = %m
// With a probability of 60%: %0 is a new allocation
%0 = test.probabilistic_alloc {p = 0.4 : f32} %m : memref<?xf32>
```

Such an op would have to declare an `Allocate` side effect. I think side effects are a "maybe" property. False positives are allowed in `getEffects`. I guess this is somewhat of an edge case. But we could support such cases correctly by requiring that ops with an `Allocate` side effect to implement the `BufferDeallocationOpInterface`, which can then be queried to build the ownership indicator.

(In the example above, without further knowledge about the op, `BufferDeallocationOpInterface::materializeUniqueOwnershipForMemref` could compare the pointers of `%0` and `%m`. If they are the same, we take the ownership indicator of `%m`. Otherwise, `%true`.)

@maerhart Does this make sense?

matthias-springer commented:
> Another tricky case would be ops that sometimes allocate and sometimes don't. This is something that we do not support properly at the moment.
> 
> ```mlir
> // With a probability of 40%: %0 = %m
> // With a probability of 60%: %0 is a new allocation
> %0 = test.probabilistic_alloc {p = 0.4 : f32} %m : memref<?xf32>
> ```
> 
> Such an op would have to declare an `Allocate` side effect. I think side effects are a "maybe" property. False positives are allowed in `getEffects`. I guess this is somewhat of an edge case. But we could support such cases correctly by requiring that ops with an `Allocate` side effect to implement the `BufferDeallocationOpInterface`, which can then be queried to build the ownership indicator.
> 
> (In the example above, without further knowledge about the op, `BufferDeallocationOpInterface::materializeUniqueOwnershipForMemref` could compare the pointers of `%0` and `%m`. If they are the same, we take the ownership indicator of `%m`. Otherwise, `%true`.)
> 
> @maerhart Does this make sense?

I am going to prepare a change that adds "maybe side effects", wich should address the ""may allocate" issue: https://discourse.llvm.org/t/how-to-model-maybe-side-effects/75674. (This may take a while...)

matthias-springer committed:
[mlir][bufferization] Buffer deallocation: Disallow unregistered ops

Memory side effects of unregistered ops are unknown. In particular, we do not know whether an unregistered op allocates memory or not. Therefore, unregistered ops cannot be handled safely in the buffer deallocation pass.

aartbik commented on mlir/lib/Dialect/Bufferization/Transforms/OwnershipBasedBufferDeallocation.cpp:
```diff
@@ -48,6 +48,32 @@ static Value buildBoolValue(OpBuilder &builder, Location loc, bool value) {
 
 static bool isMemref(Value v) { return v.getType().isa<BaseMemRefType>(); }
 
+/// Return "true" if the given op is guaranteed to have no "Allocate" or "Free"
```
This is super nit picky but the not before an or is always ambiguous (as in !Allocate or Free or !(Allocate or Free). I think you can avoid this by just saying to have neither Allocate nor Free ....

(and yes, I realize this is very nitpicky)

aartbik commented on mlir/lib/Dialect/Bufferization/Transforms/OwnershipBasedBufferDeallocation.cpp:
```diff
@@ -48,6 +48,32 @@ static Value buildBoolValue(OpBuilder &builder, Location loc, bool value) {
+           hasEffect<MemoryEffects::Free>(op);
+  // If the op does not implement the MemoryEffectOpInterface but has has
+  // recursive memory effects, then this op in isolation (without its body) does
+  // not have any side effects. The ops inside the regions of this op will be
```
All the ops inside the regions ....
(reads a bit better?)

aartbik commented on mlir/lib/Dialect/Bufferization/Transforms/OwnershipBasedBufferDeallocation.cpp:
```diff
@@ -491,7 +502,29 @@ BufferDeallocation::verifyFunctionPreconditions(FunctionOpInterface op) {
+  // do not know whether an unregistered op allocates memory or not. Call ops
+  // typically do not implement the MemoryEffectOpInterface but usually do not
+  // have side effects (apart from the callee, which will be analyzed
+  // separately), so they are allowed.
```
It would read better if you move the "are allowed" up in the sentence (so that it does not dangle after the exception given in parenthesis)

aartbik commented on mlir/test/Dialect/Bufferization/Transforms/OwnershipBasedBufferDeallocation/dealloc-region-branchop-interface.mlir:
```diff
@@ -511,58 +511,49 @@ func.func @assumingOp(
   %a = memref.alloc(%arg0) : memref<?xf32>
   scf.while (%arg1 = %a, %arg2 = %a) : (memref<?xf32>, memref<?xf32>) -> (memref<?xf32>, memref<?xf32>) {
-    %0 = "test.make_condition"() : () -> i1
+    // This op has a side effect, but it's not an allocate/free side effect.
```
this reads well!

aartbik commented on mlir/test/Dialect/Bufferization/Transforms/OwnershipBasedBufferDeallocation/dealloc-region-branchop-interface.mlir:
```diff
@@ -591,7 +582,8 @@ func.func @while_two_arg(%arg0: index) {
   %a = memref.alloc(%arg0) : memref<?xf32>
   scf.while (%arg1 = %a, %arg2 = %a, %arg3 = %a) : (memref<?xf32>, memref<?xf32>, memref<?xf32>) -> (memref<?xf32>, memref<?xf32>, memref<?xf32>) {
-    %0 = "test.make_condition"() : () -> i1
+    // This op has a side effect but it's not an Allocate or Free side effect.
```
so do the same: not an allocate/free (or otherwise the neither nor)
